srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1220 @@
+	/// be returned to notify of this. This is different behaviour to `unreserve`.
+	pub fn reserve(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Result {
+		// Do we need to consider that this is an atomic transaction?
+		let original_reserve_balance = Self::reserved_balance(asset_id, who);
+		let original_free_balance = Self::free_balance(asset_id, who);
+		if original_free_balance < amount {
+			return Err("not enough free funds");
+		}
+		let new_reserve_balance = original_reserve_balance + amount;
+		Self::set_reserved_balance(asset_id, who, new_reserve_balance);
+		let new_free_balance = original_free_balance - amount;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		Ok(())
+	}
+
+	/// Move up to `amount` from reserved balance to free balance. This function cannot fail.
+	///
+	/// much funds up to `amount` will be deducted as possible, `remaining` will be returned
+	/// NOTE: This is different to `reserve`.
+	pub fn unreserve(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> T::Balance {
+		let b = Self::reserved_balance(asset_id, who);
+		let actual = rstd::cmp::min(b, amount);
+		let original_free_balance = Self::free_balance(asset_id, who);
+		let new_free_balance = original_free_balance + actual;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		Self::set_reserved_balance(asset_id, who, b - actual);
+		amount - actual
+	}
+
+	/// Deduct up to `amount` from the combined balance of `who`, preferring to deduct from the
+	/// free balance. This function cannot fail.
+	///
+	/// much funds up to `amount` will be deducted as possible. If this is less than `amount`
+	/// then `Some(remaining)` will be returned. Full completion is given by `None`.
+	pub fn slash(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Option<T::Balance> {
+		let free_balance = Self::free_balance(asset_id, who);
+		let free_slash = rstd::cmp::min(free_balance, amount);
+		let new_free_balance = free_balance - free_slash;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		if free_slash < amount {
+			Self::slash_reserved(asset_id, who, amount - free_slash)
+		} else {
+			None
+		}
+	}
+
+	/// Adds up to `amount` to the free balance of `who`.
+	pub fn reward(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Result {
+		let original_free_balance = Self::free_balance(asset_id, who);
+		let new_free_balance = original_free_balance + amount;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		<TotalIssuance<T>>::mutate(asset_id, |x| *x = x.saturating_add(amount));
+		Ok(())
+	}
+
+	/// Deduct up to `amount` from reserved balance of `who`. This function cannot fail.
+	///
+	/// much funds up to `amount` will be deducted as possible. If this is less than `amount`
+	/// then `Some(remaining)` will be returned. Full completion is given by `None`.
+	pub fn slash_reserved(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Option<T::Balance> {
```

Comment:
This doesn't do anything with total issuance. I think the docs should mention that this function will reduce a balance without updating total issuance and it's up to the caller to handle.

Reply:
I updated to use the same docs from `Currency` trait